File: backend/app/schemas/practice_session.py
from datetime import datetime
from pydantic import BaseModel, Field


class PracticeSessionItemCreate(BaseModel):
    question_id: int
    knowledge_point_id: int | None = None
    user_answer: str | None = None
    is_correct: bool
    duration_seconds: int = Field(default=0, ge=0)


class PracticeSessionCreate(BaseModel):
    mode: str = Field(default="normal", max_length=50)
    title: str | None = Field(default=None, max_length=200)
    knowledge_base_id: int | None = None
    items: list[PracticeSessionItemCreate]
    duration_seconds: int = Field(default=0, ge=0)
    started_at: datetime | None = None
    ended_at: datetime | None = None


class PracticeSessionItemResponse(BaseModel):
    id: int
    question_id: int
    knowledge_point_id: int | None
    user_answer: str | None
    is_correct: bool
    duration_seconds: int
    question_type: str | None = None
    stem: str | None = None
    correct_answer: str | None = None
    reference_answer: str | None = None
    analysis: str | None = None
    knowledge_point_title: str | None = None
    created_at: datetime


class PracticeSessionResponse(BaseModel):
    id: int
    title: str | None
    mode: str
    knowledge_base_id: int | None
    knowledge_base_name: str | None = None
    total_count: int
    correct_count: int
    wrong_count: int
    accuracy_rate: float
    duration_seconds: int
    knowledge_point_ids: list[int]
    weak_knowledge_point_ids: list[int]
    wrong_question_ids: list[int]
    suggestion: str | None
    started_at: datetime | None
    ended_at: datetime | None
    created_at: datetime
    items: list[PracticeSessionItemResponse] | None = None


class PracticeSessionListResponse(BaseModel):
    items: list[PracticeSessionResponse]
    total: int
    page: int
    page_size: int

File: frontend/src/api/practiceSession.ts
import request from './request'

export interface PracticeSessionItemPayload {
  question_id: number
  knowledge_point_id?: number
  user_answer?: string
  is_correct: boolean
  duration_seconds?: number
}

export interface PracticeSessionPayload {
  mode?: string
  title?: string
  knowledge_base_id?: number
  items: PracticeSessionItemPayload[]
  duration_seconds?: number
  started_at?: string
  ended_at?: string
}

export interface PracticeSessionItem {
  id: number
  question_id: number
  knowledge_point_id: number | null
  user_answer: string | null
  is_correct: boolean
  duration_seconds: number
  question_type: string | null
  stem: string | null
  correct_answer: string | null
  reference_answer: string | null
  analysis: string | null
  knowledge_point_title: string | null
  created_at: string
}

export interface PracticeSession {
  id: number
  title: string | null
  mode: string
  knowledge_base_id: number | null
  knowledge_base_name: string | null
  total_count: number
  correct_count: number
  wrong_count: number
  accuracy_rate: number
  duration_seconds: number
  knowledge_point_ids: number[]
  weak_knowledge_point_ids: number[]
  wrong_question_ids: number[]
  suggestion: string | null
  started_at: string | null
  ended_at: string | null
  created_at: string
  items?: PracticeSessionItem[]
}

export interface PracticeSessionListResponse {
  items: PracticeSession[]
  total: number
  page: number
  page_size: number
}

export function createPracticeSession(data: PracticeSessionPayload) {
  return request.post('/api/practice/sessions', data) as Promise<PracticeSession>
}

export function getPracticeSessions(params?: {
  page?: number
  page_size?: number
  mode?: string
}) {
  const qs = new URLSearchParams()
  if (params?.page) qs.set('page', String(params.page))
  if (params?.page_size) qs.set('page_size', String(params.page_size))
  if (params?.mode) qs.set('mode', params.mode)
  const suffix = qs.toString() ? `?${qs.toString()}` : ''
  return request.get(`/api/practice/sessions${suffix}`) as Promise<PracticeSessionListResponse>
}

export function getPracticeSession(id: number) {
  return request.get(`/api/practice/sessions/${id}`) as Promise<PracticeSession>
}
